Random map generator expansion.

MistakeNot4892 opened the pull request:
This is a refactor of a halfassed implementation. It breaks the ore map and cave map into separate objects within a /datum/random_map class, standardizes a lot of procs, and adds a maze class and multiple terrain generators, including replace-all-space variants. 

It does NOT include the xenoruins that I was intending to include when I started the PR, as they were a bit fiddlier than intended.

This PR also remaps the mining asteroid to cover the entire Z-level and renames/refluffs it as a mining base on one of Erebus' moons. I'll tidy the lore up before a final commit; you can't reach it from space anymore though.

This PR also adds a Z-level base turf setup to avoid venting the mining base because somehow space turfs appear under shuttles.

PICS:
Maze:
http://i.imgur.com/NvKOY0b.png
Generic noise map:
http://i.imgur.com/XuGb5Sh.png
Ore distribution map:
http://i.imgur.com/EzKlfpB.png
Applied noise map (countryside):
http://i.imgur.com/62r9VNG.png
Caves:
http://i.imgur.com/g9RnT46.png

COOL PICS:
Tundra map replacing space on the derelict Z-level:
http://i.imgur.com/e03Q1L6.png
Desert map outside Arrivals:
http://i.imgur.com/XbcaaVr.png

TODO:
- [x] Add create/display/apply verbs for maps.
- [x] Refactor/split up the map generator folder.
- [x] Recalibrate the cave generator constants to produce a usable map.
- [x] Fix diamond-square map population so distribution maps generate again.

GinjaNinja32 commented on code/__HELPERS/maths.dm:
```diff
@@ -131,3 +131,14 @@ var/list/sqrtTable = list(1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4,
 /proc/norm(x, y)
 	return sqrt(squaredNorm(x, y))
+
+/proc/IsPowerOfTwo(var/val)
```
A faster version that works so long as the value is within the 16-bit limit:

```DM
/proc/IsPowerOfTwo(val)
    return (val & (val-1)) == 0
```

mwerezak replied:
Would this work?
```
/proc/IsPowerOfTwo(val)
    return (val & (~1)) == 0
```
Basically we just want to test that the least significant bit is unset.

Zuhayr replied:
Fixed it up according to GN's suggestion. As it stands, a value that exceeds 16 bits would easily overflow the map array limit.

GinjaNinja32 replied:
@mwerezak Power of two != multiple of two.

mwerezak replied:
Right. In that case you would want to test that only one bit is set.

GinjaNinja32 commented on code/__HELPERS/maths.dm:
```diff
@@ -131,3 +131,12 @@ var/list/sqrtTable = list(1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4,
+	var/check_val = 2
+	while(check_val < val)
+		check_val *= 2
+	return check_val
```
It would not at all surprise me to find that the following executes faster:

```DM
/proc/RoundUpToPowerOfTwo(val)
    return 2 ** -round(-log(2,val))
```
(it would be `2 ** ceil(log(2,val))`, but BYOND doesn't have `ceil()`)

PsiOmegaDelta commented on code/game/atoms_movable.dm:
```diff
@@ -248,14 +248,14 @@
 
 //This list contains the z-level numbers which can be accessed via space travel and the percentile chances to get there.
-var/list/accessible_z_levels = list("1" = 5, "3" = 10, "4" = 15, "5" = 10, "6" = 60)
+var/list/accessible_z_levels = list("1" = 5, "3" = 10, "4" = 15, "6" = 60)
```
Hmm, making a note to self mostly: Make config setting

PsiOmegaDelta commented on code/modules/random_map/building/facility.dm:
```diff
@@ -0,0 +1,138 @@
+
+/datum/random_map/building/facility/generate_map()
+
+	world << "build_rooms called"
```
World?

Zuhayr replied:
facility.dm is unticked, it's nonworking code.

PsiOmegaDelta replied:
Ah, easy to miss in a diff.

PsiOmegaDelta commented on code/modules/random_map/building/facility.dm:
```diff
@@ -0,0 +1,138 @@
+				if(within_bounds(current_cell))
+					map[current_cell] = ROOM_TEMP_CHAR
+
+	world << "generated [rooms.len] rooms"
```
Also here.

PsiOmegaDelta commented on code/modules/random_map/building/facility.dm:
```diff
@@ -0,0 +1,138 @@
+		var/datum/room/room = pick(rooms)
+		rooms -= room
+
+		world << "building [room.id]"
```
And here.

MistakeNot4892 committed:
Refactored random map generator system and added several terrain generators.
Created a global list to track base turfs for explosions/shuttle moves.
Remaps the asteroid to be a moonlet. Tidies up some references to 'asteroid', removes moonbase from the accessible z level list.

PsiOmegaDelta commented:
Saw no other concerns as such. It's a fairly big diff to go over on a 5 inch screen though.
Prepared to :+1: still.

PsiOmegaDelta commented:
More commits just as I +1? :,(

Zuhayr commented:
Squashed the commits down and removed the nonworking files with the world <<s in them.
